bin/propolis-server/src/lib/serial/mod.rs

```diff
@@ -2,400 +2,533 @@
+    pub async fn stop(self) {
+        let tasks = {
+            let mut guard = self.client_tasks.lock().unwrap();
+            std::mem::take(&mut guard.tasks)
+        };
 
-    loop {
-        let (uart_read, ws_send) =
-            if ws_sinks.is_empty() || cur_output.is_none() {
-                (serial.read_source(&mut output).fuse(), Fuse::terminated())
+        let mut tasks: Vec<_> = tasks.into_values().collect();
+        for task in &mut tasks {
+            let (tx, _rx) = oneshot::channel();
+            let done_tx = std::mem::replace(&mut task.done_tx, tx);
+            let _ = done_tx.send(());
+        }
+
+        futures::future::join_all(tasks.into_iter().map(|task| task.hdl)).await;
+    }
+
+    /// Creates a new task that connects a websocket stream to this manager's
+    /// serial console.
+    pub async fn connect(
+        &self,
+        ws: WebSocketStream<WebsocketConnectionRaw>,
+        kind: ClientKind,
+    ) {
+        const SERIAL_CHANNEL_SIZE: usize = 256;
+        let (console_tx, console_rx) = mpsc::channel(SERIAL_CHANNEL_SIZE);
+        let (control_tx, control_rx) = mpsc::channel(1);
+        let (task_start_tx, task_start_rx) = oneshot::channel();
+        let (task_done_tx, task_done_rx) = oneshot::channel();
+        let (permissions, discipline) = match kind {
+            ClientKind::ReadWrite => (
+                backend::Permissions::ReadWrite,
+                backend::FullReadChannelDiscipline::Block,
+            ),
+            ClientKind::ReadOnly => (
+                backend::Permissions::ReadOnly,
+                backend::FullReadChannelDiscipline::Close,
+            ),
+        };
+
+        // If this client is to be the read-write client, and there's already
+        // such a client, the old client needs to be evicted and stopped before
+        // the new client can start. Create and install the new task under the
+        // lock, then drop the lock before running down the old task.
+        let prev_rw_task;
+        {
+            let mut client_tasks = self.client_tasks.lock().unwrap();
+            let client_id = client_tasks.next_id();
+            prev_rw_task = if kind == ClientKind::ReadWrite {
+                client_tasks.remove_rw_client()
             } else {
-                let range = cur_output.clone().unwrap();
+                None
+            };
+
+            let backend_hdl =
+                self.backend.attach_client(console_tx, permissions, discipline);
+
+            let ctx = SerialTaskContext {
+                log: self.log.clone(),
+                ws,
+                backend_hdl,
+                console_rx,
+                control_rx,
+                start_rx: task_start_rx,
+                done_rx: task_done_rx,
+                client_tasks: self.client_tasks.clone(),
+                client_id,
+            };
+
+            let task = ClientTask {
+                hdl: tokio::spawn(async move { serial_task(ctx).await }),
```

Comment:
style nit: calling an async function, such as `serial_task`, returns a `Future`, so the `async move { ....await }` is unnecessary; this could just be: 
```suggestion
                hdl: tokio::spawn(serial_task(ctx)),
```